x/stake/keeper/slash.go

```diff
@@ -4,56 +4,176 @@ import (
+// so all unbonding delegations and redelegations from that height are stored
+func (k Keeper) Slash(ctx sdk.Context, pubkey crypto.PubKey, height int64, power int64, fraction sdk.Rat) {
+
+	// Amount of slashing = slash fraction * power at time of infraction
+	slashAmount := sdk.NewRat(power).Mul(fraction)
+	// hmm, https://github.com/cosmos/cosmos-sdk/issues/1348
 
-	// TODO height ignored for now, see https://github.com/cosmos/cosmos-sdk/pull/1011#issuecomment-390253957
 	validator, found := k.GetValidatorByPubKey(ctx, pubkey)
 	if !found {
-		panic(fmt.Errorf("Attempted to slash a nonexistent validator with address %s", pubkey.Address()))
+		panic(fmt.Errorf("attempted to slash a nonexistent validator with address %s", pubkey.Address()))
 	}
-	sharesToRemove := validator.PoolShares.Amount.Mul(fraction)
+	ownerAddress := validator.GetOwner()
+
+	// Track remaining slash amount
+	remainingSlashAmount := slashAmount
+
+	// Get the current pool so we can update it as we go
 	pool := k.GetPool(ctx)
-	validator, pool, burned := validator.RemovePoolShares(pool, sharesToRemove)
-	k.SetPool(ctx, pool)              // update the pool
-	k.UpdateValidator(ctx, validator) // update the validator, possibly kicking it out
 
+	if height > ctx.BlockHeight() {
+
+		// Can't slash infractions in the future
+		panic(fmt.Sprintf("impossible attempt to slash future infraction at height %d but we are at height %d", height, ctx.BlockHeight()))
+
+	} else if height == ctx.BlockHeight() {
+
+		// Special-case slash at current height for efficiency - we don't need to look through unbonding delegations or redelegations
+
+	} else if height < ctx.BlockHeight() {
+
+		// Iterate through unbonding delegations from slashed validator
+		unbondingDelegations := k.GetUnbondingDelegationsFromValidator(ctx, ownerAddress)
+		for _, unbondingDelegation := range unbondingDelegations {
+			amountSlashed := k.slashUnbondingDelegation(ctx, unbondingDelegation, height, fraction)
+			remainingSlashAmount = remainingSlashAmount.Sub(amountSlashed)
+			// Burn unbonding tokens
+			pool.LooseTokens -= amountSlashed.EvaluateInt().Int64()
+		}
+
+		// Iterate through redelegations from slashed validator
+		redelegations := k.GetRedelegationsFromValidator(ctx, ownerAddress)
+		for _, redelegation := range redelegations {
+			amountSlashed := k.slashRedelegation(ctx, redelegation, height, fraction)
+			remainingSlashAmount = remainingSlashAmount.Sub(amountSlashed)
+			// Burn unbonding shares
+			pool.UnbondingShares = pool.UnbondingShares.Sub(amountSlashed)
+		}
+
+	}
+
+	sharesToRemove := remainingSlashAmount
+	// Cannot decrease balance below zero
+	if sharesToRemove.GT(validator.PoolShares.Amount) {
+		sharesToRemove = validator.PoolShares.Amount
+	}
+
+	validator, pool, burned := validator.RemovePoolShares(pool, sharesToRemove) // remove shares from the validator
+	pool.LooseTokens -= burned                                                  // burn tokens
```

Comment:
This will be simpler once the pool share system is revmoed (https://github.com/cosmos/cosmos-sdk/pull/1119#issuecomment-400110984).